### PR TITLE
Peterc/unify counters

### PR DESCRIPTION
**Pull request by ylc015**
Rebuild timer from scratch.

background.js
only one timer implementation is used in the whole app, and it lives in background.js.
- it tracks all internal state changes for the app. 
- it periodically broadcast changes to foreground, such as popup count down text and popup view changes
- update badge text, this can only be done in background because we need to make sure the timer still works even if user is not interacting with our app.

popup.js
the controller of timers. is aware of the focus/break timer. User mostly interact with this window since they have to click on start/stop button.  
- calls background timer API to start/stop timers
- listen for state changes in background so that it could update the count-down text and popup view 

option.js
- send duration values as soon as user updates the timer durations

how initialization works.
there are two ways to initialize the timer in background.
1. popup.js. user will always click on it so we do the timer initialization here. And we get the focus/break timer duration from local storage ( written by option.js ).
2. option.js, user doesn't have to interact with this page at all if they are just using defaults. but when they do update the durations, the change is immediately propagated to background and the timer *will be reset* to reflect new duration

notes:
You might notice delays in UI changes. that is because the update interval in background is set to one seconds. For optimal user experience, we might need to update every 100ms. Not sure how much stress that will put on the whole extensions performance.

### CODE REVIEW

File: src/js/background.js [hunk 1]
```diff
@@ -1,62 +1,232 @@
 "use strict";
-
-// Initialized default timers
-chrome.runtime.onInstalled.addListener(function() {
-  chrome.storage.sync.set({
-    view: "focusView",
-    breakTime: 5,
-    focusTime: 30,
-    // Can combine timerStarted and startTime into one variable
```
Comment (leannezhang) [inline src/js/background.js:9]: we can use `const` for public constants

File: src/js/background.js [hunk 2]
```diff
@@ -1,62 +1,232 @@
...
-  // When the focus alarm times up
-  if (alarms.name === "focusAlarm") {
-    // seems like when the windows is not closed, the notification bar would not appear
-    chrome.storage.sync.set(
-      { view: "breakView", timerStarted: false, startTime: 0 },
-      function() {
-        console.log("The view is breakView");
-      }
-    );
+function setExerciseURL (url) {
```
Comment (leannezhang) [inline src/js/background.js]: I don't think this function is used.

File: src/js/background.js [hunk 3]
```diff
@@ -1,62 +1,232 @@
...
-    );
+  } else if (type === exitingBreakView) {
     chrome.notifications.create({
       title: "Break is over",
       message: "Break is over. Time to focus now! Click on the timer to start.",
       iconUrl: "src/images/get_started128.png",
       type: "basic"
     });
+  } else if (type === toFocusView) {
+    // toFocusView
```
Comment (leannezhang) [inline src/js/background.js]: can remove this comment

File: src/js/background.js [hunk 4]
```diff
@@ -1,62 +1,232 @@
...
+    console.error(`Caught unexpected type ${type}`)
   }
-  chrome.browserAction.setBadgeText({ text: "" });
-});
+}
+
+
+/***************************************************************** */
+// Focus/Break Timer API
+//timer functions
```
Comment (leannezhang) [inline src/js/background.js]: can remove the "timer function" comment duplicated with "focus/break timer api" comment

File: src/js/background.js [hunk 5]
```diff
@@ -1,62 +1,232 @@
...
+var running = 'running'
+var stopped = 'stopped'
+var unInit = 'unInit'
+
+var focusTimeInitialDurationMs = -1;
+var breakTimeInitialDurationMs = -1;
+
+var unInitView = 'unInitView';
+var inFocusView = 'inFocusView';
+var inBreakView = 'inBreakView';
```
Comment (leannezhang) [inline src/js/background.js:103]: we can use const

File: src/js/background.js [hunk 6]
```diff
@@ -1,62 +1,232 @@
...
+var viewState = unInitView;
```
Comment (leannezhang) [inline src/js/background.js:104]: can use let here

File: src/js/background.js [hunk 7]
```diff
@@ -1,62 +1,232 @@
...
+
+// could be one of
+// running || stopped 
+var timerState = unInit;
+var running = 'running'
+var stopped = 'stopped'
+var unInit = 'unInit'
+
+var focusTimeInitialDurationMs = -1;
+var breakTimeInitialDurationMs = -1;
```
Comment (leannezhang) [inline src/js/background.js:99]: why -1 not 0?

File: src/js/background.js [hunk 8]
```diff
@@ -1,62 +1,232 @@
...
+
+/***************************************************************** */
+// Timer API
+
+// could be one of
+// running || stopped 
+var timerState = unInit;
+var running = 'running'
+var stopped = 'stopped'
+var unInit = 'unInit'
```
Comment (leannezhang) [inline src/js/background.js]: what does unInit state looks like? Does it mean it's "not-started"?

File: src/js/background.js [hunk 9]
```diff
@@ -1,62 +1,232 @@
...
+
```
Comment (leannezhang) [inline src/js/background.js:97]: can we move `let timerState = unInit;` behind "running" & "stopped" and "unInit"

File: src/js/background.js [hunk 10]
```diff
@@ -1,62 +1,232 @@
...
+var breakTimeInitialDurationMs = -1;
+
+var unInitView = 'unInitView';
+var inFocusView = 'inFocusView';
+var inBreakView = 'inBreakView';
+var viewState = unInitView;
+
+// this should match the default in option
+// as there are no ways to sync the default value from option.html
+let _initialDurationMs = 30 * 60 * 1000;
```
Comment (leannezhang) [inline src/js/background.js]: is this the default focus time? 30 mins? perhaps we can put the rename this to defaultDurationMs.

thanks for the comment.

File: src/js/background.js (hunk 8 above)
Comment (leannezhang) [reply src/js/background.js]: unitialized?

File: src/js/background.js (hunk 9 above)
Comment (leannezhang) [reply src/js/background.js:97]: will timerState only gets set to uninitialized once in the beginning before the timer started? Did this ever gets reset again?

File: src/js/background.js (hunk 10 above)
Comment (leannezhang) [reply src/js/background.js]: defaultDurations = 30 * 60 * 1000;
durationsMs = defaultDurations;

File: src/js/background.js (hunk 8 above)
Comment (leannezhang) [reply src/js/background.js]: changed to "uninitialized"

File: src/js/background.js [hunk 11]
```diff
@@ -1,62 +1,232 @@
...
+  resetTimer()
+}
+
+function startTimer() {
+  console.log(`starting timer with duration ${_initialDurationMs}`)
+  _remainingMs = _initialDurationMs
+  timerState = running;
+};
+
+// duplicate?
```
Comment (leannezhang) [inline src/js/background.js]: remove this comment

**Review by leannezhang — COMMENTED** [review]
Good job!!!!!!!

File: src/js/background.js (hunk 6 above)
Comment (ylc015) [reply src/js/background.js:104]: cannot use either let or var here because they are not globally available

File: src/js/background.js (hunk 5 above)
Comment (ylc015) [reply src/js/background.js:103]: cannot use either let or var here because they are not globally available

**Pull request merged**